hls4ml/model/optimizer/passes/infer_precision.py

```diff
@@ -0,0 +1,290 @@
+
+        return ['result_t']
+
+    def _infer_common_precision(self, node, types_to_infer, n_ops):
+        inferred_types = []
+
+        input_precision = node.get_input_variable().type.precision
+        input_width = input_precision.width
+        input_integers = input_precision.integer
+
+        if 'weight_t' in types_to_infer:
+            weight_quantizer = node.get_attr('weight_quantizer', None)
+            if weight_quantizer is not None:
+                weight_width = weight_quantizer.bits
+                weight_integers = weight_quantizer.hls_type.integer
+                node.types['weight_t'].name = node.name + '_weight_t'
+                node.types['weight_t'].precision = weight_quantizer.hls_type
+            else:
+                self._infer_default_type(node, 'weight_t')
+                weight_width = node.types['weight_t'].precision.width
+                weight_integers = node.types['weight_t'].precision.integer
+            node.weights['weight'].update_precision(node.types['weight_t'].precision)
+
+            inferred_types.append('weight_t')
+        else:
+            weight_width = node.types['weight_t'].precision.width
+            weight_integers = node.types['weight_t'].precision.integer
+
+        if 'bias_t' in types_to_infer:
+            bias_quantizer = node.get_attr('bias_quantizer', None)
+            if bias_quantizer is not None:
+                bias_width = bias_quantizer.bits
+                bias_integers = bias_quantizer.hls_type.integer
+                node.types['bias_t'].name = node.name + '_bias_t'
+                node.types['bias_t'].precision = bias_quantizer.hls_type
+            else:
+                self._infer_default_type(node, 'bias_t')
+                bias_width = node.types['bias_t'].precision.width
+                bias_integers = node.types['bias_t'].precision.integer
+            node.weights['bias'].update_precision(node.types['bias_t'].precision)
+
+            inferred_types.append('bias_t')
+        else:
+            bias_width = node.types['bias_t'].precision.width
+            bias_integers = node.types['bias_t'].precision.integer
+
+        new_type = FixedPrecisionType(
+            width=int(max(np.ceil(input_width + weight_width + np.log2(n_ops)), bias_width) + 1),
+            integer=int(max(np.ceil(input_integers + weight_integers + np.log2(n_ops)), bias_integers) + 1),
+        )
+
+        if 'accum_t' in types_to_infer:
+            node.types['accum_t'].name = node.name + '_accum_t'
+            node.types['accum_t'].precision = new_type
+
+            inferred_types.append('accum_t')
+
+        if 'result_t' in types_to_infer:
+            node.types['result_t'].name = node.name + '_result_t'
+            node.types['result_t'].precision = new_type
+
+            inferred_types.append('result_t')
+
+        return inferred_types
+
+    def _infer_dense_precision(self, node, types_to_infer):
+        n_ops = node.get_attr('n_in') * node.get_attr('n_out')
```

Comment:
I don't think the total n_ops is the important value for the accumulator precision. Ignoring bias for now, each output value is a certain number of multiplies, resulting in the `input_width + weight_width` part of the equation, plus the accumulation, `math.ceil(np.log2(num_acc))`. But num_acc != num_ops. In particular, it's `node.get_attr('n_in')` (-1?), at least for the standard 1D Dense layer. Bias modifies things a bit, but the general trends are the same. This is the result we had from the CMS hackathon with Sioni:  https://github.com/jmitrevs/hls4ml/blob/bit-correct/hls4ml/model/optimizer/passes/propagate_dense_precision.py

Reply:
Ah, true, I'll fix this.